### DISCUSSION TITLE
ImageMagick Portable - RegistryKeyLookupFailed

### DISCUSSION THREAD

**Discussion by DadeMurphy72** [ImageMagick/ImageMagick#7434] [Help] [answered]
### ImageMagick version

7.1.1-34-portable-Q16-x64

### Operating system, version and so on

Windows 10, 11

### Description

I've built an application in Python and it does exactly what I need to do. When I compile it in INNO setup, I continue to get a RegistryKeyLookupFailed error when trying to convert an image. So I built a small test app to try to isolate the issue. I am stuck. I've tried everything I can think of. Even in this test app, when it installs, it tells me where each path is and it's set correctly. I also auto generate the delegates file when the app is ran.

My test script is:
```

import os
import sys
import tkinter as tk
from tkinter import filedialog, messagebox
from wand.image import Image as WandImage

def create_delegates_file(delegates_path, gs_path):
    delegates_content = f'''<?xml version="1.0"?>
<delegatemap>
   <delegate decode="eps" command="&quot;{gs_path}&quot; -q -dQUIET -dSAFER -dBATCH -dNOPAUSE -dNOPROMPT -sstdout=%stderr -sDEVICE=pngalpha -dTextAlphaBits=4 -dGraphicsAlphaBits=4 -r%[resolution.x]x%[resolution.y] -dUseCIEColor -sOutputFile=%o -f %i -c showpage -c quit"/>
   <delegate decode="ai" command="&quot;{gs_path}&quot; -q -dQUIET -dSAFER -dBATCH -dNOPAUSE -dNOPROMPT -sstdout=%stderr -sDEVICE=pngalpha -dTextAlphaBits=4 -dGraphicsAlphaBits=4 -r%[resolution.x]x%[resolution.y] -dUseCIEColor -sOutputFile=%o -f %i -c showpage -c quit"/>
</delegatemap>'''
    with open(delegates_path, 'w') as file:
        file.write(delegates_content)

def convert_file(input_path, output_path):
    try:
        with WandImage(filename=input_path) as img:
            img.format = 'png'
            img.save(filename=output_path)
            print(f'Converted {input_path} to {output_path}')
    except Exception as e:
        print(f'Failed to convert file: {e}')

def main():
    # Debugging output
    print(f'MAGICK_HOME: {os.environ.get("MAGICK_HOME")}')
    print(f'PATH: {os.environ.get("PATH")}')
    print(f'MAGICK_CODER_MODULE_PATH: {os.environ.get("MAGICK_CODER_MODULE_PATH")}')
    print(f'MAGICK_CONFIGURE_PATH: {os.environ.get("MAGICK_CONFIGURE_PATH")}')

    magick_home = os.environ.get("MAGICK_HOME")
    gs_path = os.path.join(magick_home, "ghostscript/bin/gswin64.exe")
    delegates_path = os.path.join(magick_home, "delegates.xml")

    create_delegates_file(delegates_path, gs_path)
    print(f'Delegates file created at {delegates_path}')

    root = tk.Tk()
    root.withdraw()
    input_path = filedialog.askopenfilename(title="Select AI or EPS file", filetypes=[("AI and EPS files", "*.ai *.eps")])
    if input_path:
        output_path = filedialog.asksaveasfilename(title="Save as PNG", defaultextension=".png", filetypes=[("PNG files", "*.png")])
        if output_path:
            convert_file(input_path, output_path)
    else:
        messagebox.showerror("Error", "No file selected")

if __name__ == "__main__":
    main()

```
For my iss, I've tried multiple things and even forcing it to run as an adminstrator:
```
#define MyAppName "Convert AI/EPS to PNG"
#define MyAppVersion "1.0"

[Setup]
AppName={#MyAppName}
AppVersion={#MyAppVersion}
DefaultDirName={commonpf}\ConvertAItoPNG
DefaultGroupName={#MyAppName}
OutputDir=D:\Projects\TEST\dist\installer
OutputBaseFilename=ConvertAItoPNGSetup
PrivilegesRequired=admin
ArchitecturesAllowed=x64
ArchitecturesInstallIn64BitMode=x64
Compression=lzma2
SolidCompression=yes
SetupIconFile=D:\Projects\TEST\images\PJC.ico

[Files]
Source: "D:\Projects\TEST\dist\convert_ai_eps_to_png\*"; DestDir: "{app}"; Flags: ignoreversion recursesubdirs createallsubdirs

[Icons]
Name: "{group}\Convert AI/EPS to PNG"; Filename: "{app}\convert_ai_eps_to_png.exe"
Name: "{group}\Uninstall Convert AI/EPS to PNG"; Filename: "{uninstallexe}"

[Run]
Filename: "{app}\convert_ai_eps_to_png.exe"; Description: "Run Convert AI/EPS to PNG"; Flags: nowait postinstall skipifsilent

[Code]
function SetEnvironmentVariable(lpName, lpValue: String): Boolean;
external 'SetEnvironmentVariableA@kernel32.dll stdcall';

procedure InitializeEnvironmentVariables;
var
  MAGICK_HOME, MAGICK_CODER_MODULE_PATH, MAGICK_CONFIGURE_PATH: String;
begin
  MAGICK_HOME := ExpandConstant('{app}\resources\ImageMagick');
  MAGICK_CODER_MODULE_PATH := MAGICK_HOME + '\modules\coders';
  MAGICK_CONFIGURE_PATH := MAGICK_HOME;

  // Set environment variables
  if not SetEnvironmentVariable('MAGICK_HOME', MAGICK_HOME) then
    MsgBox('Error setting MAGICK_HOME', mbError, MB_OK);
  if not SetEnvironmentVariable('MAGICK_CODER_MODULE_PATH', MAGICK_CODER_MODULE_PATH) then
    MsgBox('Error setting MAGICK_CODER_MODULE_PATH', mbError, MB_OK);
  if not SetEnvironmentVariable('MAGICK_CONFIGURE_PATH', MAGICK_CONFIGURE_PATH) then
    MsgBox('Error setting MAGICK_CONFIGURE_PATH', mbError, MB_OK);

  // Displaying the set environment variables for debugging purposes
  MsgBox('MAGICK_HOME: ' + MAGICK_HOME, mbInformation, MB_OK);
  MsgBox('MAGICK_CODER_MODULE_PATH: ' + MAGICK_CODER_MODULE_PATH, mbInformation, MB_OK);
  MsgBox('MAGICK_CONFIGURE_PATH: ' + MAGICK_CONFIGURE_PATH, mbInformation, MB_OK);
end;

procedure CurStepChanged(CurStep: TSetupStep);
begin
  if CurStep = ssInstall then
  begin
    InitializeEnvironmentVariables;
  end;
end;
```

I've been working on this all week and have searched many times over. I'm at a loss and I'm sure I'm missing something simple. Any help is appreciated!

**Comment by dlemstra** [comment]
You should not be in the part of the code that throws this exception with a portable build. Are you sure that you are using the portable build?

**Reply by DadeMurphy72** [reply]
Thanks for replying. I am sure I have the portable download. I had questioned it since it was asking for a registry key so I downloaded it again and copied the files over. I’ll download again and try it all over.

**Comment by DadeMurphy72** [comment]
Just tested every x64 portable version, all with the same results. Is there an earlier version I can try? Maybe there was an error in uploading this build?

**Reply by dlemstra** [reply]
I am not able to reproduce your issue in a clean VM with the portable version. I suspect this must be something on your side.

**Reply by DadeMurphy72** [reply]
Very strange, I'm installing this on a test computer (2 different ones) and I've checked my (system) environment variables, I have nothing set for ImageMagick. The downloads I have are the portable versions from the main site. I expand and compile. I've marked each download and all have the same results. I understand you are saying it would be something on my side, and I'm not disagreeing with that, but what else should I be checking? I do not have ImageMagick or Python installed. I have everything running through my app folder. I ran this as an administrator through a cmd prompt window to see if there was any additional information, but there wasn't. 

Out of curiosity, did you download 7.1.1-34-portable-Q16-x64 and try it this way?

<img width="466" alt="image" src="https://github.com/ImageMagick/ImageMagick/assets/4086083/e65d4e37-3935-49b6-b0c8-a6accd40bcc3">

**Reply by dlemstra** [reply]
I downloaded the portable executable from our website and executed a conversion in a clean VM. I did not get any error message. It really looks like you are using a non portable version of ImageMagick. The portable build is a static build and should not hit this code path. I don't think there is much else I can do to help you.

**Comment by DadeMurphy72** [comment, marked as answer]
I installed the correct portable version. I compiled and installed on a new machine and it was showing me the non-portable version. I checked my spec file and it was pointing to the correct portable version but for whatever reason, continued to include the installed version on my system. I then changed the spec file for the explicit location and was able to get the correct version installed. Thanks.

**Reply by dlemstra** [reply]
Happy to hear that you resolved it and thanks for coming back to explain how you fixed it.
